### ISSUE TITLE
Compare large dataset; compare result options?

### ISSUE THREAD

**Issue by Unforgettable631** [six42/jdbcslim#7]
I think it's not an issue, more a question about using the compare functionality.

**Situation**
I've setup a test to compare the outcome (source vs target). Ideally the test result doesn't find any differences. In our case we're not ready yet and a few differences will occur. The outcome is a large dataset (3000+ rows) on two columns. At the moment in the outcome there are only 5 differences, but on the testpage there are more then 1000 + differences which is causing confusion. I think I'm missing an ouputparameter or so to determine how the resultsheets must be compared? Which options are there for result-comparison?

I hope the question is clear. Details about the testpage and a grep of the testresult added below.
I expect in the grep only two differences in the final resultsheet, row 2 and 5:
Values:
[1],[b] and [2],[f]

Here's how the compare page looks like:

```

!|Define Properties|Source                     |
|key               |value                      |
|jdbcDriver        |SQLDriver                  |
|DBURL             |${connection_url}|
|DBUSER            |user                       |
|.keyStoreLocation |keystore                   |
|DBPASSWORD        |password                   |
|#DBAUTOCOMMIT     |true                       |
|CMD               |${source_query}            |
|query             |true                       |

!|Define Properties|Target                     |
|key               |value                      |
|jdbcDriver        |SQLDriver                  |
|DBURL             |${connection_url}|
|DBUSER            |user                       |
|.keyStoreLocation |keystore                   |
|DBPASSWORD        |password                   |
|#DBAUTOCOMMIT     |true                       |
|CMD               |${target_query}            |
|query             |true                       |


!|Script   |SQLCommand|Source|
|open connection             |
|execute                     |
|$RSSource=|resultSheet      |
|close connection            |

!|Script   |SQLCommand  |Target   |
|open connection                  |
|execute                          |
|$RSTarget=|resultSheet           |
|$RSDiff=  |compareSheet|$RSSource|
|close connection                 |


!|Table:SheetEcho|$RSDiff|

```

Grep of the sheetresults and compare flags:

```
_Source result_
[TABLE_NAME],[COLUMN_NAME]
[1],[a]
[1],[b]
[2],[c]
[2],[d]
[2],[e]

_Target result_
[TABLE_NAME],[COLUMN_NAME]
[1],[a]
[2],[c]
[2],[d]
[2],[e]
[2],[f]


_Compare_
[report:TABLE_NAME, report:COLUMN_NAME],
[pass:1, pass:a],
[fail:[2] expected[1], fail:[c] expected[b]],
[pass:2, fail:[d] expected [c]],
[pass:2, fail:[e] expected [d]],
[pass:2, fail:[f] expected [e]]

```

**Comment by six42** [comment]
A very good question. And a tricky edge case.
Thanks for providing clear details about your scenario.

If I understand you correctly you would like to see in your example that  [1],[b] has been removed and [2],[f] has been added.

To allow jdbcSlim to identify additions and removals  the key columns in your data must be defined: See UserGuide.6OutputOptions.Sorting how to define keys.

So in the table with the expected data the key columns must be flagged. In your case this is the source and I think both columns are keys. 
As you don't enter the expected values with the header row but generate them with an SQL statement your statement must return the header with the key information.

In short change your source_query to :  
select TABLE_NAME as "TABLE_NAME<", COLUMN_NAME as "COLUMN_NAME<" from ...

Your target query must not be changed and must not contain the "<" signs!

make sure your data is returned by your SQL in sorted order. 
Then it should work. :)

**Comment by Unforgettable631** [comment]
Thanks for your quick reply. The solution you provided for defining the keys is exactly what I was looking for. Thanks a lot!

I was already close to the solution. I tried the solution yesterday on the target query, but this fails in a Java error. It isn't exactly clear (without any knowledge of the jdbcSlim code) the keys can be defined like this. I will make a specific example to clarify how this works if it also helps for other users (a sort of addition of the documentation).

**Comment by six42** [comment]
Glad that I could help you.

A PR is always welcome.
Also FitNesse makes writing good documentation  simpler, still some features  are missed.
